### ISSUE TITLE
meta charset="UTF-8"

### ISSUE THREAD

**Issue by ronyeh** [vexflow/vexflow#140]
I've noticed in some demos, I get weird symbols if I don't explicitly declare the charset to be **UTF-8**.

```html
  <head>
    <meta charset="UTF-8" />
  </head>
```

Without the declaration, I get this:

<img width="539" alt="image" src="https://github.com/vexflow/vexflow/assets/239113/3e39f408-0cdd-4cd4-8728-840e795466c9">

With the declaration, it looks better:

<img width="531" alt="image" src="https://github.com/vexflow/vexflow/assets/239113/c1c37c0e-595d-45bf-8891-11a23970c018">

This makes sense, since we are now displaying text instead of rendering paths. It is just something to note. Maybe we need to put these little potential stumbling blocks into a developer's FAQ at some point.

**Comment by 0xfe** [comment]
This is a bummer. Is there a way to do this programatically from the library JS itself?

As a stopgap, this should be on the front page of the docs, not just the FAQ. Otherwise, it will immediately put off first time devs using the library.

**Comment by ronyeh** [comment]
Yes, we shouldn't have to deal with this problem in this day and age.

....I have a guess as to why I only just ran into this issue tonight. (I haven't seen this issue since we started working on V5.)

I was updating really old JSFiddle examples (e.g. sandbox), and it looks like these pages start with this:

```
<!DOCTYPE >
```

However, the correct modern way to start an HTML5 page is:
```
<!DOCTYPE html>
```

UTF-8 is the default character set in HTML5, so I believe this should not be a big concern, as long as people start their pages with the correct doctype.

So I will go ahead and update our *really* old examples to fix the doctype, and hopefully we won't run into this again.

I can still add it to a FAQ (and link to this issue).

**Comment by ronyeh** [comment]
I investigated and conclude that yes, this is a potential stumbling block that we need to mention in all our examples and docs.


If the <meta charset='UTF-8'> is not specified, the browser will try to autodetect the character set (if the web server did not specify the charset in the Content-Type header).

When I test from the file system (by dragging a HTML file to the browser), there is no web server so the browser needs to auto-detect the character set.

The jsfiddle demo pages are empty, so Chrome defaults to the 'windows-1252' encoding. Verify this by opening the Chrome console and typing: `document.characterSet`

When I test the same file (with no charset declaration) from `npx http-server`, it works fine because that web server sends a header `Content-Type: text/html; charset=UTF-8` by default.

If I add UTF-8 text to the page, for example the GBP symbol £, then Chrome will autodetect UTF-8 and the score will render properly.

In summary, the problem pops up if all of the below are true:
- web server does not send charset=UTF-8, OR the HTML file is loaded from file://
- HTML file does NOT declare `<meta charset='UTF-8'>`
- HTML file does NOT contain obvious UTF-8 content, which would help the browser auto-detect UTF-8.

**Comment by ronyeh** [comment]
More info:

https://www.w3.org/TR/2011/WD-html5-20110405/semantics.html#charset


> Authors are encouraged to use UTF-8. Conformance checkers may advise authors against using legacy encodings. [[RFC3629]](https://www.w3.org/TR/2011/WD-html5-20110405/references.html#refsRFC3629)
> 
> Authoring tools should default to using UTF-8 for newly-created documents. [[RFC3629]](https://www.w3.org/TR/2011/WD-html5-20110405/references.html#refsRFC3629)
> 


So I don't think there is a default charset for HTML 5. Everyone is encouraged to use UTF-8, but clearly Chrome doesn't choose UTF-8 by default when it has no information about the document's character set.

**Comment by ronyeh** [comment]
Here's an informative page:

https://hsivonen.fi/label-utf-8/

> ...always encode your HTML as UTF-8. Furthermore, in order to let browsers know that the document is UTF-8-encoded, always label it as such. To label your document, you need to do at least one of the following:
> 
> * Put \<meta charset="utf-8"> as the first thing after the \<head> start tag (i.e. as the first child of head).
> 
> The meta tag, including its ending > character needs to be within the first 1024 bytes of the file. Putting it right after the \<head> start tag is the easiest way to get this right. Do not put comments before the \<head> start tag and \<meta charset="utf-8"> to avoid accidentally pushing the latter past the first 1024 bytes.
> 
> * Configure your server to send the header Content-Type: text/html; charset=utf-8 on the HTTP layer.
> 
> * Start the document with the UTF-8 BOM, i.e. the bytes 0xEF, 0xBB, and 0xBF.
> 
> Doing more than one of these is OK.

**Comment by 0xfe** [comment]
Okay, so no real solution than "make it loud and clear!"

**Comment by ronyeh** [comment]
I don't think there is a way to change the charset dynamically in JS (or specify it in the SVG output).

Unfortunately the only "solution" right now is to make sure all our examples include the charset declaration, and we have to mention it at the top of the README and repeat it multiple times in other documentation. :-(

@rvilarl have you run into this issue?

**Comment by rvilarl** [comment]
No I have not but I had not headers in vexflow-examples. Now they are in.

**Comment by sschmidTU** [comment]
I had an idea for an automatic fix / hack: setting the meta charset attribute programmatically.
If vexflow does this early on, it might help.
It might be a hack or not the ideal way for things to go, but if it works and doesn't cause issues, why not?

Here's what I tried with a document that does declare `<meta charset="utf-8">` right after `<head>`:
<img width="364" alt="image" src="https://github.com/vexflow/vexflow/assets/33069673/2fe48df3-fd2a-454c-b0c3-5e92281546ba">
```js
document.getElementsByTagName('meta')[0].getAttribute("charset")
document.getElementsByTagName('meta')[0].setAttribute("charset", "utf-16") // of course, you'd set it to utf-8
```

The most compatible procedure may be:
* Check if there's a meta element with a valid charset attribute. If it exists, don't change it (might e.g. cause issues with utf-16).
* If there is none, add a new meta element with charset utf-8. (Try to make it the first child of `<head>`)

Of course, it might be that changing this programmatically doesn't help, because it's too late or otherwise.
Also, it does seem to be recommended for developers to set the charset attribute themselves anyways (just saying).

This is a bit of work to experiment with, so, don't feel pressured to do it immediately, just thought I'd put this idea here.
